### PR TITLE
Early PR: Initial Commit of POM for review

### PR DESCRIPTION
**Pull request by b-w-cole**

### CODE REVIEW

**Comment by b-w-cole** [comment]
This commit contains POM refactoring for all test cases, except for webTables.  Could you review what I have so far?

File: pages/ownerInformationPage.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    }
+
+    async clickAddPet(){
+        
+        await this.page.getByRole('button', {name: 'Add New Pet'}).click()
+
+    }
+
+    async editPet(petName: string){
```
Comment (ba-instructor) [inline pages/ownerInformationPage.ts]: It needs a better name.
For example, the method above you have called "clickAddPet" which clearly explains the action.
But "editPet" is not clear... does it mean that this method updates Pet Name? Or maybe other information?
So make it clear, like "clickEditPetButtonFor". There will be no double meaning for sure.

File: pages/ownerInformationPage.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            expect(await row.textContent()).not.toContain(formattedDate)
+        }
+    }
+
+    async validateVisitDateDisplayForTopVisit(petName: string, appointmentDate: Date){
+        const formatedDate = this.getDateDashFormat(appointmentDate)
+        await expect(this.page.locator('app-pet-list', {hasText: petName}).locator('app-visit-list tr td').first()).toHaveText(formatedDate)
+    }
+
+    async compareOrderOfTopTwoVisitsByDate(petName: string){
```
Comment (ba-instructor) [inline pages/ownerInformationPage.ts]: Not a descriptive naming as well.
What is the condition of "comparison"?
It would be better to do something like "validateTwoLastVisitsInDesendOrder"

File: pages/ownersPage.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import { Page, expect } from '@playwright/test';
+import { HelperBase } from './helperBase';
+
+export class OwnersPage extends HelperBase{
+
+    constructor (page : Page){
+        super(page)
+    }
+
+    async getOwnerRow(owner : string){
```
Comment (ba-instructor) [inline pages/ownersPage.ts:10]: It looks like this method is never used

File: pages/ownersPage.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        await this.page.waitForSelector('tbody')
+        console.log(await this.page.getByRole('row', {name: owner}).allTextContents())
+        return this.page.getByRole('row', {name: owner})
+    }
+
+    async selectOwner(owner: string){
```
Comment (ba-instructor) [inline pages/ownersPage.ts]: Not descriptive enough, how owner can be selected.
So the name something like "selectOwnerByName" and the parameter 'ownerName" would be better

File: pages/ownersPage.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        await this.page.getByText(owner).click()
+    }
+
+    async validatePetNameAndCityForOwner(petName: string, petOwner: string, city: string){
```
Comment (ba-instructor) [inline pages/ownersPage.ts]: This method was never used

File: pages/ownersPage.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    
+        // Validate the owner is from Monona
+        await expect(row.locator('td').nth(2)).toHaveText('Monona')
+    
+        // Validate the owner has a pet named Lucky
+        await expect(row.locator('td').nth(4)).toHaveText('Lucky')
+        
+    }
+
+    async validatePetNumberPerCity(city: string){
```
Comment (ba-instructor) [inline pages/ownersPage.ts]: This method was never used

File: pages/petDetailsPage.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+    async validateNameOwnerAndPetTypeFieldsDisplayCorrectly(petName: string, owner: string, petType: string){
+        // Validate that the name input field is correct
+        await expect(this.page.locator('#name')).toHaveValue(petName)
+        await expect(this.page.locator('#owner_name')).toHaveValue(owner)
+        await expect(this.page.locator('#type1')).toHaveValue(petType)
+
+    }
+
+    async validateAllPetTypesDisplayCorrectly(){
```
Comment (ba-instructor) [inline pages/petDetailsPage.ts]: Again, naming is confusing.
Where pet types should be displayed correctly? What "correctly" even mean?

What this method is doing it "validateAllPetValuesFromDropdownDisplayedInPetTypeField"

File: pages/petDetailsPage.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+            await this.page.selectOption('select', petType)
+            
+            // Validate that the pet is correctly displayed in the textbox
+            await expect(this.page.locator('#type1')).toHaveValue(petType)
+        }
+    }
+
+
+    async selectNewPetTypeAndUpdate(petType: string){
```
Comment (ba-instructor) [inline pages/petDetailsPage.ts]: How about "updatePetTypeTo"

File: pages/petTypesPage.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        /* 5. Add a new pet type with the name "pig" and click "Save" button */
+        await input.click()
+        await input.fill(petType)
+        await this.page.getByText('Save').click()
+
+        // Validate that the new Pet Type has been added
+        await expect(this.page.locator('table input').last()).toHaveValue(petType)
+    }
+
+    async editPetType(oldPetType: string){
```
Comment (ba-instructor) [inline pages/petTypesPage.ts]: a better name would be "clickEditPetTypeButton" as more descriptive

File: pages/specialtiesPage.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+                await row.getByText('Delete').click()
+                break;
+            }
+        }
+
+    }
+
+    async editSpecialty(specialty : string){
+
+    }
```
Comment (ba-instructor) [inline pages/specialtiesPage.ts]: Empty method

File: pages/veterinariansPage.ts [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import { Page } from '@playwright/test';
+import { HelperBase } from './helperBase';
+
+export class VeterinariansPage extends HelperBase{
+    constructor(page : Page){
+        super(page)
+    }
+    
+    async editVeterinarian(name: string){
```
Comment (ba-instructor) [inline pages/veterinariansPage.ts]: This method is only clicks "Edit Vet" button, make the method name to reflect it

**Review by ba-instructor — COMMENTED** [review]
Please check the comments and refer to the commends I have shared during our 101.
Mainly need to adjust the naming of the page objects to make tests clearly readable about the steps that have to be performed.

File: pages/petDetailsPage.ts (hunk 7 above)
Comment (b-w-cole) [reply pages/petDetailsPage.ts]: Updated this page object to be EditPetPage

File: pages/petDetailsPage.ts (hunk 8 above)
Comment (b-w-cole) [reply pages/petDetailsPage.ts]: Updated method name and moved to EditPetPage

File: package.json [hunk 12]
```diff
@@ -8,8 +8,11 @@
   "author": "",
   "license": "ISC",
   "devDependencies": {
-    "@playwright/test": "1.39.0",
+    "@playwright/test": "^1.47.2",
     "@types/node": "^20.9.3",
     "dotenv": "^16.3.1"
+  },
+  "dependencies": {
+    "playwright": "^1.46.1"
```
Comment (ba-instructor) [inline package.json]: Why did you add a playwright to "dependencies" and why this version is lower then in "devDependencies".
It think this import is not needed

File: pages/editPetPage.ts [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        await expect(this.page.locator('#name')).toHaveValue(petName)
+        await expect(this.page.locator('#owner_name')).toHaveValue(owner)
+        await expect(this.page.locator('#type1')).toHaveValue(petType)
+
+    }
+
+    async validateAllPetValuesFromDropdownDisplayedInPetTypeField(){
+
+        // Iterate through each pet type option to validate that the pet is correctly listed in the textbox to the left 
+        const allPetTypeOptions = await this.page.locator("option").allTextContents()
```
Comment (ba-instructor) [inline pages/editPetPage.ts:21]: This method can be potentially false postitve.
If `allPetTypeOptions` will be empty, the for loop below will not be executed so assertion simply will be skipped.
Currently this method is working because test case has a previous step with assertions `validateNameOwnerAndPetTypeFieldsDisplayCorrectly()` and because of that, page is properly loaded. But if you remove this step from the test case, you will see how the loop not executed.

File: pages/ownersPage.ts [hunk 14]
```diff
@@ -13,22 +13,76 @@ export class OwnersPage extends HelperBase{
...
+            // Add the pet name to the comparing list of names, with each name trimmed of excess white space
+            actualPetsList.push( (await row.getByRole('cell').nth(4).textContent())?.trim()! )
+        }
+    
+        // Comparing both sorted lists for equality
+        expect (actualPetsList.sort()).toEqual(petList.sort())
+
+    }
+
+    async searchForOwnerByLastName(ownerLastName: string){
```
Comment (ba-instructor) [inline pages/ownersPage.ts]: This name does not reflect the result. Search owner and do what? :)

File: tests/inputFields.spec.ts [hunk 15]
```diff
@@ -8,93 +8,61 @@ test.beforeEach( async({page}) => {
...
-  /* 6. Click on "Update" button */
-  // We will try to update the field anyway
-  await pm.onPetTypeEditPage().updateNewPetType()
+  await pm.onPetTypeEditPage().fillPetType('')
 
-  /* 7. Add assertion that "Edit Pet Type" page is still displayed */
-  // Verifying that we are still on the same Edit page
-  await pm.onPetTypeEditPage().validatePageHeader('Edit Pet Type')
+  await pm.onPetTypeEditPage().updatePetType()
+  await pm.onPetTypeEditPage().updatePetType()
```
Comment (ba-instructor) [inline tests/inputFields.spec.ts]: Why do you call method updatePetType() two times?
Also the implementation of this method is confusing, because this method also checks the error message when not message is typed into the input field. It would be better to separate the scope of clicking Update button and validation of error message.

File: pages/specialtiesPage.ts [hunk 16]
```diff
@@ -7,12 +7,41 @@ export class SpecialtiesPage extends HelperBase{
...
 
+    async validateExistenceOf(specialty: string){
+        const entries = await this.page.getByRole('textbox').all()
+
+        for(let entry of entries){
+            if(specialty === await entry.inputValue()){
+                return true
+            }
+        }
+        return false
```
Comment (ba-instructor) [inline pages/specialtiesPage.ts]: This is not a good implementation because when a test fails, you will see something like "Expected true but actual false" in the error message. Not descriptive at all. Also in the test level it also look not very nice, when you pass the page object method as argument into the expect assertion.
Just make this validation inside of this method and make method be responsible for the validation. It will make test looks cleaner and assertion error message more meaningful when test will fail.
P.S. you can use `toContainEqual()` assertion to validate if your array of values has expected value or not

**Review by ba-instructor — COMMENTED** [review]
Great updates!
Please check a few comments.

File: package.json (hunk 12 above)
Comment (b-w-cole) [reply package.json]: This was an accident, kind of.  I was struggling with my VSCode and reinstalled/updated, which probably added this.  I can revert it.  You're right, it's not necessary.

File: pages/ownersPage.ts (hunk 14 above)
Comment (b-w-cole) [reply pages/ownersPage.ts]: Hmm, ok, I am also checking for the existence of owners, so I have come up with `searchForOwnerByLastNameAndValidateOwnerExistsFor(ownerLastName: string)`

File: tests/inputFields.spec.ts (hunk 15 above)
Comment (b-w-cole) [reply tests/inputFields.spec.ts]: Hmmm...  I had a reason for it, but I can no longer remember!  >.<   Edited this comment - it didn't format correctly, either.  I'm going to put it in Slack instead.

File: pages/editPetPage.ts (hunk 13 above)
Comment (b-w-cole) [reply pages/editPetPage.ts:21]: Good catch!  I'm thinking about this one... I'm going to ask my question in Slack.

File: tests/inputFields.spec.ts (hunk 15 above)
Comment (b-w-cole) [reply tests/inputFields.spec.ts]: Actually, I think I got it and am going to go ahead and submit it when ready.  I tested it and it works... confirmed that the 'Name is required' message assertion is picked up.

```
async updatePetType(){
      const petTypeInputField = this.page.getByRole('textbox')
      await this.page.getByText('Update').click()
      if(await petTypeInputField.inputValue() === ''){
          await expect(this.page.locator('div', {hasText: 'Name'}).last()).toHaveText('Name is required')
      }
  }
  ```

File: pages/specialtiesPage.ts (hunk 16 above)
Comment (b-w-cole) [reply pages/specialtiesPage.ts]: Thanks for this feedback!  It's super helpful.  I was able to re-work a lot of it, but have some remaining questions that I will post in Slack channel.

File: pages/petTypeEditPage.ts [hunk 17]
```diff
@@ -21,11 +21,11 @@ export class PetTypeEditPage extends HelperBase{
     async updatePetType(){
         const petTypeInputField = this.page.getByRole('textbox')
 
+        await this.page.getByText('Update').click()
+
         if(await petTypeInputField.inputValue() === ''){
             await expect(this.page.locator('div', {hasText: 'Name'}).last()).toHaveText('Name is required')
         }
-        
-        await this.page.getByText('Update').click()
```
Comment (ba-instructor) [inline pages/petTypeEditPage.ts]: I have mentioned it in my previous comment, but I ll add it one more time.
The concern with this method I have is that it combines two things:
1. Clicking of Update button
2. Validation of the message "Name is required" **WHEN** the input field is empty.

By reading this test, tell me, is it clear where validation should happen if you don't know steps implementation?
image

**Review by ba-instructor — COMMENTED** [review]
Looks good )
Check one comment )

File: pages/petTypeEditPage.ts (hunk 17 above)
Comment (b-w-cole) [reply pages/petTypeEditPage.ts]: Oh sorry, I may have overlooked this.  I can definitely see how this is unclear.  I was trying to keep it simple, but it is ultimately not.  One minute and I'll have an update.

**Review by ba-instructor — APPROVED** [review]
